src/core/thread/mle_router.cpp

```diff
@@ -1425,27 +1424,7 @@ Error MleRouter::HandleAdvertisement(RxInfo &aRxInfo)
             IgnoreError(SendLinkRequest(nullptr));
         }
 
-        router = mRouterTable.FindRouterById(routerId);
-        VerifyOrExit(router != nullptr);
-
-        // check current active router number
-        routerCount = 0;
-
-        for (uint8_t id = 0; id <= kMaxRouterId; id++)
-        {
-            if (routeTlv.IsRouterIdSet(id))
-            {
-                routerCount++;
-            }
-        }
```

Comment:
Adding a note on the change here:
- In the exitsing code, we count the routers in `routeTlv` we get from neighboring router (number of routers from its perspective) and then check `routerCount > mRouterDowngradeThreshold`.
- In the new code in `ShouldDowngrade()`, we now use `mRouterTable.GetActiveRouterCount()` (number of routers from our perspective).

I am not sure if it matters in most cases, but wanted to mention this change. My interpretation of reading spec is that it should be our router count.

Reply:
I think doing it based on local route table works for this case. We want to trigger the downgrade evaluation whenever we process a new Route TLV, which this does.